sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+    if (body) {
+      if (isReadableStream(body)) {
+        request.headers.set("Transfer-Encoding", "chunked");
+      } else if (typeof body === "string") {
+        request.headers.set("Content-Length", body.length);
+      }
+    }
+
+    try {
+      const result = await new Promise<PipelineResponse>((resolve, reject) => {
+        const options = getRequestOptions(request);
+        const req = https.request(options, async (res) => {
+          const headers = getResponseHeaders(res);
+
+          const status = res.statusCode ?? 0;
+          const response: PipelineResponse = {
+            status,
+            headers,
+            request
+          };
+
+          let responseStream = getResponseStream(res, headers, request.skipDecompressResponse);
+
+          const onDownloadProgress = request.onDownloadProgress;
+          if (onDownloadProgress) {
+            const downloadReportStream = new ReportTransform(onDownloadProgress);
+            responseStream.pipe(downloadReportStream);
+            responseStream = downloadReportStream;
+          }
+
+          if (request.streamResponseBody) {
+            response.readableStreamBody = responseStream;
+          } else {
+            response.bodyAsText = await streamToText(responseStream);
+          }
+
+          resolve(response);
+        });
+        req.on("error", (err) => {
+          reject(new RestError(err.message, { code: RestError.REQUEST_SEND_ERROR, request }));
+        });
+        abortController.signal.addEventListener("abort", () => {
+          if (!req.finished) {
+            req.abort();
+            reject(new AbortError("The operation was aborted."));
+          }
+        });
+        if (body) {
+          if (isReadableStream(body)) {
+            body.pipe(req);
+          } else {
+            req.write(body);
+          }
+        }
+        req.end();
+      });
+      return result;
+    } finally {
+      // clean up event listener
+      if (request.abortSignal && abortListener) {
+        request.abortSignal.removeEventListener("abort", abortListener);
+      }
+    }
+  }
+}
+
+function prepareFormData(formData: FormDataMap, request: PipelineRequest): void {
+  const requestForm = new FormData();
+  for (const formKey of Object.keys(formData)) {
+    const formValue = formData[formKey];
+    if (Array.isArray(formValue)) {
+      for (const subValue of formValue) {
+        requestForm.append(formKey, subValue);
+      }
+    } else {
+      requestForm.append(formKey, formValue);
+    }
+  }
+
+  request.body = requestForm;
+  request.formData = undefined;
+  const contentType = request.headers.get("Content-Type");
+  if (contentType && contentType.indexOf("multipart/form-data") !== -1) {
+    request.headers.set(
+      "Content-Type",
+      `multipart/form-data; boundary=${requestForm.getBoundary()}`
+    );
+  }
+}
+
+function getRequestOptions(request: PipelineRequest): https.RequestOptions {
+  const agent = getOrCreateAgent(request);
```

Comment:
That seems reasonable. Are you thinking we'd add a node-only requestoption for this or something else?

Reply:
Yes exactly, a node-only request option.

Reply:
Hmm are we worried about having to type `agent` in requestOptions? It seems awfully big

Reply:
well maybe not big, but it's a class rather than an interface

```ts
export class Agent {
        maxFreeSockets: number;
        maxSockets: number;
        sockets: any;
        requests: any;

        constructor(opts?: AgentOptions);

        /**
         * Destroy any sockets that are currently in use by the agent.
         * It is usually not necessary to do this. However, if you are using an agent with KeepAlive enabled,
         * then it is best to explicitly shut down the agent when you know that it will no longer be used. Otherwise,
         * sockets may hang open for quite a long time before the server terminates them.
         */
        destroy(): void;
    }
```

Is it worth creating a compatible interface to that or should we do something like unknown/any?

Reply:
We talked about this briefly today, seems like we may be able to use placeholder types eventually, but for now this can be unknown/any

Reply:
Is there a way to pass in our own HTTP agent or to specify max sockets today? I'm aiming to allow the use of both node-fetch and Azure Search SDK in an Azure Function to make requests to `<mysearch>.search.windows.net`, but need to ensure I don't run into port exhaustion.

Sorry for dredging up an old topic, but this is the closest discussion I've found.

Reply:
@mdmower please always feel free to open a new issue even if it's just to ask a question. We're happy to help and filing something helps to ensure that your feedback/inquiry won't get lost in the shuffle.

To answer your specific question, I think you could do this by creating a custom `PipelinePolicy` that sets the `agent` property on each `PipelineRequest` made by the Search SDK when performing an operation. To have the SDK use your custom policy, you can pass it via the `additionalPolicies` client option.

If you have trouble getting it to work, feel free to file an issue! 😄

Reply:
Thanks @xirzec, I really appreciate the reply. I haven't figured it out yet, so I posted a new [issue](https://github.com/Azure/azure-sdk-for-js/issues/25522).